crates/subspace-farmer/src/bin/subspace-farmer/commands/shared/network.rs

```diff
@@ -120,17 +133,79 @@ where
         allow_non_global_addresses_in_dht: allow_private_ips,
         known_peers_registry,
         request_response_protocols: vec![
-            PieceByIndexRequestHandler::create(move |_, &PieceByIndexRequest { piece_index }| {
+            {
+                let maybe_weak_node = Arc::clone(&maybe_weak_node);
+                let farmer_cache = farmer_cache.clone();
+
+                CachedPieceByIndexRequestHandler::create(move |peer_id, request| {
+                    let CachedPieceByIndexRequest {
+                        piece_index,
+                        mut cached_pieces,
+                    } = request;
+                    debug!(?piece_index, "Cached piece request received");
+
+                    let maybe_weak_node = Arc::clone(&maybe_weak_node);
+                    let farmer_cache = farmer_cache.clone();
+
+                    async move {
+                        let piece_from_cache =
+                            farmer_cache.get_piece(piece_index.to_multihash()).await;
+                        cached_pieces.truncate(MAX_CACHED_PIECES);
+                        let cached_pieces = farmer_cache.has_pieces(cached_pieces).await;
+
+                        Some(CachedPieceByIndexResponse {
+                            result: match piece_from_cache {
+                                Some(piece) => PieceResult::Piece(piece),
+                                None => {
+                                    let maybe_node = maybe_weak_node
+                                        .lock()
+                                        .as_ref()
+                                        .expect("Always called after network instantiation; qed")
+                                        .upgrade();
+
+                                    let closest_peers = if let Some(node) = maybe_node {
+                                        node.get_closest_local_peers(
```

Comment:
I think this is an incorrect p2p design:
- a simple request from the source peer results in multiple requests on the remote peer. It's a limited form of the "amplification attack"
- a request timeout can contradict the Kademlia timeout resulting in an unpredictable   outcome

Reply:
I think there is some misunderstanding here, `get_closest_local_peers` doesn't make any network requests, it is a quick look into local Kademlia view calling https://github.com/libp2p/rust-libp2p/blob/9a45db3f82b760c93099e66ec77a7a772d1f6cd3/protocols/kad/src/behaviour.rs#L1215-L1229 internally

Reply:
Yes, I incorrectly treated `get_closest_local_peers` as a distributed operation with a similar name. @teor2345 @nazar-pc 

I don't have objections against a local in-memory query.